#1: Record all steps

sampraxis opened the pull request:
Fix step history doesn't diff correctly

Improve assignment submission data

Extract majority of business logic from locallib.php into it own class

Change (Authorization) setting to (OpenAI API token) instead

Improve user experience

sampraxis committed:
#1: Record all steps

Fix step history doesn't diff correctly

Improve assignment submission data

Extract majority of business logic from locallib.php into it own class

Change (Authorization) setting to (OpenAI API token) instead

Improve user experience

frederikmillingpytlick commented on classes/app/ai/history/archive.php:
```diff
@@ -17,109 +13,163 @@
-    private moodle_transaction $transaction;
-    private ?attempt_entity $attempt = null;
+    private string $type;
+    private ?int $submission;
```
More specific naming

```suggestion
    private ?int $submission_id;
```

frederikmillingpytlick commented on classes/app/file/repository.php:
```diff
@@ -0,0 +1,103 @@
+        return 'submissions_pxaiwriter';
+    }
+
+    public function get_submission_files(context $context, object $submission): array
```
You are only using the id?

```suggestion
    public function get_submission_files(context $context, int $submission_id): array
```

frederikmillingpytlick commented on classes/app/helper/diff/text.php:
```diff
@@ -0,0 +1,97 @@
+        return '</span>';
+    }
+
+    private function get_granularity(?string $type): GranularityInterface
```
You are only using $this->granularity, no need for param

```suggestion
    private function get_granularity(): GranularityInterface
```

frederikmillingpytlick commented on classes/app/interfaces/factory.php:
```diff
@@ -22,11 +25,25 @@ public function assign(): assign_factory;
 
     public function ai(): ai_factory;
 
+    /**
+     * @template T
+     * @psalm-template T
+     * @param array $items
+     * @return collection<T>
+     */
```
Did you get this to work?

sampraxis replied:
Sometimes.

If you define it at the method it doesn't work probably. Somehow this feature feel like it's still in beta :sweat_smile: 
But at the ``class``, ``interface`` works ok (not perfect).

Since Jetbrains said that they support it [link to their statement](https://blog.jetbrains.com/phpstorm/2021/07/phpstorm-2021-2-release/). We might just add it for the convenient.

Maybe it work if we change PHP level to PHP 8.1+ in the setting.

frederikmillingpytlick commented on classes/app/moodle/interfaces/factory.php:
```diff
@@ -25,7 +26,23 @@ public function curl(array $settings = []): curl;
+
+    public function get_config_instance(string $component = base_factory::COMPONENT): object;
+
+    /**
+     * @template T
+     * @psalm-template T
+     * @param string $name
+     * @param T $value
+     * @param string $component
+     * @return void
+     */
```
Why this doc block?

frederikmillingpytlick commented on classes/external/ai/generate_ai_text.php:
```diff
@@ -111,12 +121,14 @@ public  static function execute(int $assignment_id, string $text, int $step = 1)
-                $error
-            );
-            throw $error;
+            try
+            {
+                $transaction->rollback($exception);
+            }
+            catch (\Exception $rollback_exception) {}
```
:laughing:

sampraxis replied:
Yeah Moodle throw an exception  when we call ``rollback``. That is very mess up...

sampraxis committed:
#1: Record all steps

Apply the changes from code review by @frederikmillingpytlick

Remove unused code

sampraxis committed:
#1: Record all steps

Add 3rd party library requirements for Moodle

sampraxis committed:
#1: Record all steps

Fix plugin cannot read history from empty submission (new)

Fix JS doesn't provide submission id for the history record

sampraxis committed:
#1: Record all steps

Disable text pasting in user input

Fix user own text shouldn't count as diff

Fix generated AI text should overwrite all other steps text input

sampraxis committed:
#1: Record all steps

Fix AI expand diff text

Fix PDF annotation diff

Fix should only show the data that had been save by user

Fix record empty history

sampraxis committed:
#1: Record all steps

Fix history table in status default value

sampraxis committed:
#1: Record all steps

Fix history record that ignore empty string
